pandas/_libs/index.pyx

```diff
@@ -585,16 +571,23 @@ cpdef convert_scalar(ndarray arr, object value):
         raise ValueError("cannot set a Timedelta with a non-timedelta "
                          f"{type(value).__name__}")
 
-    if (issubclass(arr.dtype.type, (np.integer, np.floating, np.complex)) and
-            not issubclass(arr.dtype.type, np.bool_)):
-        if util.is_bool_object(value):
-            raise ValueError("Cannot assign bool to float/integer series")
+    else:
+        validate_numeric_casting(arr.dtype, value)
+
+    return value
+
 
-    if issubclass(arr.dtype.type, (np.integer, np.bool_)):
+cpdef validate_numeric_casting(dtype, object value):
+    # Note: we can't annotate dtype as cnp.dtype because that cases dtype.type
+    #  to integer
+    if issubclass(dtype.type, (np.integer, np.bool_)):
         if util.is_float_object(value) and value != value:
             raise ValueError("Cannot assign nan to integer series")
 
-    return value
+    if (issubclass(dtype.type, (np.integer, np.floating, np.complex)) and
```

Comment:
can you not just do 

elsif (.....) as by-definition they are not bools at this point.

Reply:
the case where issubclass(dtype.type, np.integer) will pass through in both cases i think